File: macos/AICoachHotkey.m
#import <AppKit/AppKit.h>
#import <Carbon/Carbon.h>

static const OSType AICoachSignature = 'AICH';
static EventHotKeyRef AICoachHotKey = NULL;
static EventHandlerRef AICoachHandler = NULL;

static OSStatus HandleAICoachHotKey(EventHandlerCallRef nextHandler, EventRef event, void *data) {
    (void)nextHandler;
    (void)data;
    EventHotKeyID identifier = {0};
    GetEventParameter(event, kEventParamDirectObject, typeEventHotKeyID, NULL,
                      sizeof(identifier), NULL, &identifier);
    if (identifier.signature != AICoachSignature || identifier.id != 1) return noErr;

    NSTask *task = [[NSTask alloc] init];
    task.executableURL = [NSURL fileURLWithPath:@"/usr/bin/env"];
    task.arguments = @[@"aicoach", @"toggle"];
    task.standardOutput = [NSFileHandle nullDevice];
    task.standardError = [NSFileHandle nullDevice];
    [task launchAndReturnError:NULL];
    return noErr;
}

@interface AICoachAppDelegate : NSObject <NSApplicationDelegate>
@end

@implementation AICoachAppDelegate
- (void)applicationDidFinishLaunching:(NSNotification *)notification {
    (void)notification;
    EventTypeSpec type = {kEventClassKeyboard, kEventHotKeyPressed};
    OSStatus handlerStatus = InstallEventHandler(GetApplicationEventTarget(),
                                                  HandleAICoachHotKey, 1, &type,
                                                  NULL, &AICoachHandler);
    EventHotKeyID identifier = {AICoachSignature, 1};
    OSStatus hotKeyStatus = RegisterEventHotKey(kVK_Space, optionKey, identifier,
                                                GetApplicationEventTarget(), 0,
                                                &AICoachHotKey);
    if (handlerStatus != noErr || hotKeyStatus != noErr) {
        fputs("aicoach-hotkey: Option+Space registration failed (possibly already in use)\n",
              stderr);
        [NSApp terminate:nil];
    }
}

- (void)applicationWillTerminate:(NSNotification *)notification {
    (void)notification;
    if (AICoachHotKey != NULL) UnregisterEventHotKey(AICoachHotKey);
    if (AICoachHandler != NULL) RemoveEventHandler(AICoachHandler);
}
@end

int main(void) {
    @autoreleasepool {
        NSApplication *application = [NSApplication sharedApplication];
        AICoachAppDelegate *delegate = [[AICoachAppDelegate alloc] init];
        application.delegate = delegate;
        [application setActivationPolicy:NSApplicationActivationPolicyAccessory];
        [application run];
    }
    return 0;
}

File: macos/AICoachHotkey.swift
import AppKit
import Carbon
import Foundation

private let signature: OSType = 0x41494348 // "AICH"
private let hotKeyID: UInt32 = 1

final class AppDelegate: NSObject, NSApplicationDelegate {
    private var hotKey: EventHotKeyRef?
    private var handler: EventHandlerRef?

    func applicationDidFinishLaunching(_ notification: Notification) {
        var type = EventTypeSpec(eventClass: OSType(kEventClassKeyboard), eventKind: UInt32(kEventHotKeyPressed))
        InstallEventHandler(GetApplicationEventTarget(), { _, event, _ in
            var id = EventHotKeyID()
            GetEventParameter(event, EventParamName(kEventParamDirectObject), EventParamType(typeEventHotKeyID), nil, MemoryLayout<EventHotKeyID>.size, nil, &id)
            guard id.signature == signature, id.id == hotKeyID else { return noErr }
            DispatchQueue.global(qos: .userInitiated).async {
                let task = Process()
                task.executableURL = URL(fileURLWithPath: "/usr/bin/env")
                task.arguments = ["aicoach", "toggle"]
                task.standardOutput = FileHandle.nullDevice
                task.standardError = FileHandle.nullDevice
                try? task.run()
            }
            return noErr
        }, 1, &type, nil, &handler)

        var id = EventHotKeyID(signature: signature, id: hotKeyID)
        let status = RegisterEventHotKey(UInt32(kVK_Space), UInt32(optionKey), id, GetApplicationEventTarget(), 0, &hotKey)
        if status != noErr {
            fputs("aicoach-hotkey: Option+Space registration failed (possibly already in use)\n", stderr)
            NSApplication.shared.terminate(nil)
        }
    }

    func applicationWillTerminate(_ notification: Notification) {
        if let hotKey { UnregisterEventHotKey(hotKey) }
        if let handler { RemoveEventHandler(handler) }
    }
}

let app = NSApplication.shared
let delegate = AppDelegate()
app.delegate = delegate
app.setActivationPolicy(.accessory)
app.run()
